apps/voting/app/src/App.js

```diff
@@ -1,352 +1,276 @@
-      return
-    }
+  // Add some useful data to render the votes
+  const votes = useMemo(
+    () =>
+      appStateReady
+        ? appState.votes.map(vote => ({
+            ...vote,
+            data: {
+              ...vote.data,
+              open: isVoteOpen(vote, now),
 
-    this.setState({
-      userAccountVotes: new Map(
-        await Promise.all(
-          votes.map(vote =>
-            app
-              .call('getVoterState', vote.voteId, userAccount)
-              .toPromise()
-              .then(result => [vote.voteId, result])
-          )
-        )
-      ),
-    })
-  }
+              // Render text fields
+              descriptionNode: voteTextNode(vote.data.description),
+              metadataNode: voteTextNode(vote.data.metadata),
+            },
+            userAccountVote: voteTypeFromContractEnum(
+              userAccountVotes.get(vote.voteId)
+            ),
+          }))
+        : appState.votes,
+    [appStateReady, cacheIdFromVotes(appState.votes), userAccountVotes]
+  )
 
-  getTokenContract(tokenAddress) {
-    return tokenAddress && this.props.app.external(tokenAddress, tokenAbi)
-  }
-  handleCreateVote = question => {
-    this.props.app.newVote(EMPTY_CALLSCRIPT, question)
-    this.handleCreateVoteClose()
-  }
-  handleCreateVoteOpen = () => {
-    this.setState({ createVoteVisible: true })
-  }
-  handleCreateVoteClose = () => {
-    this.setState({ createVoteVisible: false })
-  }
-  handleVoteOpen = voteId => {
-    const exists = this.props.votes.some(vote => voteId === vote.voteId)
-    if (!exists) return
-    this.setState({
-      currentVoteId: voteId,
-      voteVisible: true,
-      voteSidebarOpened: false,
-    })
-  }
-  handleVote = (voteId, voteType, executesIfDecided = true) => {
-    this.props.app.vote(voteId, voteType === VOTE_YEA, executesIfDecided)
-    this.handleVoteClose()
-  }
-  handleExecute = voteId => {
-    this.props.app.executeVote(voteId)
-    this.handleVoteClose()
-  }
-  handleVoteClose = () => {
-    this.setState({ voteVisible: false })
-  }
-  handleVoteTransitionEnd = opened => {
-    this.setState(opened ? { voteSidebarOpened: true } : { currentVoteId: -1 })
-  }
+  const currentVote =
+    currentVoteId === -1
+      ? null
+      : votes.find(vote => vote.voteId === currentVoteId)
 
-  handleMenuPanelOpen = () => {
-    this.props.sendMessageToWrapper('menuPanel', true)
-  }
+  // update token contract
+  useEffect(() => {
+    setTokenContract(
+      api && appState.tokenAddress
+        ? api.external(appState.tokenAddress, tokenAbi)
+        : null
+    )
+  }, [api, appState.tokenAddress])
 
-  handleResolveLocalIdentity = address => {
-    return this.props.app.resolveAddressIdentity(address).toPromise()
-  }
-  handleShowLocalIdentityModal = address => {
-    return this.props.app
-      .requestAddressIdentityModification(address)
-      .toPromise()
-  }
+  // update user account votes
+  useEffect(() => {
```

Comment:
⚠️ Unfortunately, the caching for this effect is also prone to failure.

Since we use `cacheIdFromVotes()` again, this will only update if we add a vote. It won't be re-evaluted if the user changes their vote.

Rather than the current dependencies list, we can improve the caching and also reduce the number of network requests by:

- Applying some throttling (most useful for initial loads)
- Once we've fetched the voter status for a vote that's executed or has ended (`!isVoteOpen()`), we know, _for sure_, that you can't change your vote any more and don't need to re-evaluate it again

------------

Question about `useEffect()`: does it let us know which of the dependencies changed?

In our case, if `connectedAccount()` changes, we should start with a completely new `Map`. It's probably not a problem to recreate the `Map` each on each computation, but it'd be nice to just keep reusing the same `Map`.

Reply:
Same here, the caching has been removed.

> Question about useEffect(): does it let us know which of the dependencies changed?

It doesn’t, but there are different ways to achieve this if needed (e.g. using several `useEffect()` with `useState()`).

> In our case, if connectedAccount() changes, we should start with a completely new Map. 

It is the case yes: if either `api`, `votes` or `connectedAccount` changes, a new `Map` gets created.

> It's probably not a problem to recreate the Map each on each computation, but it'd be nice to just keep reusing the same Map.

I’m not sure I understand what you meant here? Only one `Map` gets created every time we either have an update of `connectedAccount` or `votes`, and it only contains the information needed for the connected account: `vote ID => vote status`.

Reply:
Not having to recreate the map is an absolutely tiny optimization, since we could technically reuse it each time if the connected account hasn't changed.

But recreating it each time is also good for being more immutable / React-like :).